x/shield/keeper/purchase.go

```diff
@@ -1,56 +1,79 @@
+}
+
+// AddPurchase sets a purchase of shield.
+func (k Keeper) AddPurchase(ctx sdk.Context, poolID uint64, purchaser sdk.AccAddress, purchase types.Purchase) {
+	purchaseList, found := k.GetPurchaseList(ctx, poolID, purchaser)
+	if !found {
+		purchaseList = types.NewPurchaseList(poolID, purchaser, []types.Purchase{purchase})
+	} else {
+		purchaseList.Entries = append(purchaseList.Entries, purchase)
+	}
+	k.SetPurchaseList(ctx, purchaseList)
 }
 
-// GetPurchase gets a purchase from store by txhash.
-func (k Keeper) GetPurchase(ctx sdk.Context, txhash []byte) (types.Purchase, error) {
+// GetPurchaseList gets a purchase from store by txhash.
+func (k Keeper) GetPurchaseList(ctx sdk.Context, poolID uint64, purchaser sdk.AccAddress) (types.PurchaseList, bool) {
 	store := ctx.KVStore(k.storeKey)
-	bz := store.Get(types.GetPurchaseTxHashKey(txhash))
+	bz := store.Get(types.GetPurchaseListKey(poolID, purchaser))
 	if bz != nil {
-		var purchase types.Purchase
+		var purchase types.PurchaseList
 		k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &purchase)
-		return purchase, nil
+		return purchase, true
 	}
-	return types.Purchase{}, types.ErrPurchaseNotFound
+	return types.PurchaseList{}, false
 }
 
-// DeletePurchase deletes a purchase of shield.
-func (k Keeper) DeletePurchase(ctx sdk.Context, txhash []byte) error {
+// GetPurchase gets a purchase out of a purchase list
+func GetPurchase(purchaseList types.PurchaseList, purchaseID uint64) (types.Purchase, bool) {
+	for _, entry := range purchaseList.Entries {
+		if entry.PurchaseID == purchaseID {
+			return entry, true
+		}
+	}
+	return types.Purchase{}, false
+}
+
+// DeletePurchaseList deletes a purchase of shield.
+func (k Keeper) DeletePurchaseList(ctx sdk.Context, poolID uint64, purchaser sdk.AccAddress) error {
 	store := ctx.KVStore(k.storeKey)
-	_, err := k.GetPurchase(ctx, txhash)
-	if err != nil {
-		return err
+	_, found := k.GetPurchaseList(ctx, poolID, purchaser)
+	if !found {
+		return types.ErrPurchaseNotFound
 	}
-	store.Delete(types.GetPurchaseTxHashKey(txhash))
+	store.Delete(types.GetPurchaseListKey(poolID, purchaser))
 	return nil
 }
 
 // DequeuePurchase dequeues a purchase from the purchase queue
-func (k Keeper) DequeuePurchase(ctx sdk.Context, purchase types.Purchase) {
-	timeslice := k.GetPurchaseQueueTimeSlice(ctx, purchase.ClaimPeriodEndTime)
-	for i, entry := range timeslice {
-		if bytes.Equal(entry, purchase.TxHash) {
-			timeslice = append(timeslice[:i], timeslice[i+1:]...)
-			break
+func (k Keeper) DequeuePurchase(ctx sdk.Context, purchaseList types.PurchaseList, endTime time.Time) {
+	timeslice := k.GetPurchaseQueueTimeSlice(ctx, endTime)
+	for i, ppPair := range timeslice {
+		if (purchaseList.PoolID == ppPair.PoolID) && purchaseList.Purchaser.Equals(ppPair.Purchaser) {
+			if len(timeslice) > 1 {
+				timeslice = append(timeslice[:i], timeslice[i+1:]...)
+				k.SetPurchaseQueueTimeSlice(ctx, endTime, timeslice)
+			} else {
```

Comment:
This is an opportunity of Golang style of early return to avoid unnecessary else branch.

if len() > 1 {
   ...
   return
}
...
return

Reply:
OK